### PR TITLE
Control/implement condition tree

### PR DESCRIPTION
**Pull request by MiryamBenHaim**
Implement building tree condition in a logical way and update conditons according to the sensors messeges

### CODE REVIEW

File: include/condition.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using namespace std;
+
+class Node;
+
+// Abstract class that present a node in the tree
+class Condition {
+public:
+    // status: contains what is the boolean value of this subtree
+    bool status = false;
+    // parents: contains pointers to all parents I'm their son- reduce same subtrees in the tree
```
Comment (ChedvaErblich) [inline include/condition.h]: This comment is not clear enough

File: include/enums.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    be,  // >=
+    le,  // <=
+    UnknownOperatorType
+};
+
+// Function that take a string and return the corresponding operator type
+OperatorTypes convertStringToOperatorTypes(string operatorType);
+
+// An enum that contains all possible sensor types
+enum SensorsTypes { Speed, TirePressure, Light, UnknownSensorType };
```
Comment (ChedvaErblich) [inline include/enums.h]: You can't know which sensors you'll have.
What does this enum use for?

File: include/enums.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#ifndef __ENUMS_H__
```
Comment (ChedvaErblich) [inline include/enums.h]: The name of the file is too general, choose a more specific name.
Moreover, there are other things in the file besides functions.

File: include/full_condition.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#include <unordered_map>
+
+#include "sensor.h"
+#include "or_operator.h"
+#include "and_operator.h"
+#include "basic_condition.h"
+#include "root.h"
+#include "condition_factory.h"
+
+using namespace std;
```
Comment (ChedvaErblich) [inline include/full_condition.h]: Avoid using "using namespace std" in header files

File: include/full_condition.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    int id;
+    // Root node of the condition tree
+    Root *root;
+    // Vector of actions associated with this condition
+    vector<pair<int, string>> actions;
+
+    // Constructor: Initializes the FullCondition object, parses the condition string, and builds the condition tree.
+    FullCondition(string condition, vector<pair<int, string>> &actions);
+    void activateActions();
+};
```
Comment (ChedvaErblich) [inline include/full_condition.h]: Add a blank line

File: include/global_properties.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <memory>
+
+#include "input.h"
+#include "full_condition.h"
+
+using namespace std;
```
Comment (ChedvaErblich) [inline include/global_properties.h]: Same here,
Avoid using this in header files

File: include/input.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+#include <vector>
+#include <fstream>
+#include <unordered_map>
+
+#include "sensor.h"
+#include "speed_sensor.h"
+#include "tire_pressure_sensor.h"
+#include "full_condition.h"
+
+using namespace std;
```
Comment (ChedvaErblich) [inline include/input.h]: same here

File: include/node.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#ifndef __NODE_H__
+#define __NODE_H__
+
+#include "condition.h"
+
+class Node : public Condition {
+public:
+    // countTrueConditions: contains the of children whose status is true
```
Comment (ChedvaErblich) [inline include/node.h]: Fix this comment

File: include/operator_node.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#ifndef __OPERATOR_NODE_H__
+#define __OPERATOR_NODE_H__
+
+#include <iostream>
+#include <vector>
+#include <optional>
+#include "node.h"
+using namespace std;
```
Comment (ChedvaErblich) [inline include/operator_node.h]: same here

File: include/operator_node.h [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+// Abstract class that present an internal node in the tree
+class OperatorNode : public Node {
+public:
+    // conditions: contains all the sub conditions that this condition consists of- the children of the node
+    vector<Condition *> conditions;
+
+    // Virtual D-tor to allow inheritance
```
Comment (ChedvaErblich) [inline include/operator_node.h]: I think this is clear enough and you don't need this comment.

File: include/root.h [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+public:
+    // Tree id
+    int conditionId;
+    // Pointer to the first node of the tree
+    Condition *firstCondition;
+
+    // C-tor that gets an id and initializes the conditionId
+    Root(int id, Condition *firstCondition) : conditionId(id)
+    {
+        this->firstCondition = firstCondition;
```
Comment (ChedvaErblich) [inline include/root.h]: Why you didn’t  initialize it  via the memory initialization list?

File: control/README.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# Project-in-CMake
```
Comment (ChedvaErblich) [inline control/README.md:1]: The link to sensors should be in the sensors directory.
in general, we need to think about this README, I'm not sure it's the right place for this links, but let's think about it later

File: control/include/basic_condition.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (ChedvaErblich) [inline control/include/basic_condition.h:1]: Blank line

File: control/include/condition.h [hunk 14]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (ChedvaErblich) [inline control/include/condition.h:1]: Blank line

File: control/include/global_properties.h [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <memory>
+
+#include "input.h"
+#include "full_condition.h"
+
+using namespace std;
```
Comment (ChedvaErblich) [inline control/include/global_properties.h]: Same here

File: control/include/test_helper.h [hunk 16]
```diff
@@ -0,0 +1,16 @@
+#ifndef TEST_HELPER_H
```
Comment (ChedvaErblich) [inline control/include/test_helper.h]: If it's only for testing, it should be placed in the test directory.

File: control/include/test_helper.h [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+#define TEST_HELPER_H
+
+#include <string>
+#include <iostream>
+#include <vector>
+#include "sensor.h"
+#include "input.h"
+#include "full_condition.h"
+#include "global_properties.h"
+using namespace std;
```
Comment (ChedvaErblich) [inline control/include/test_helper.h]: Same here

File: control/include/tire_pressure_sensor.h [hunk 18]
```diff
@@ -0,0 +1,15 @@
+#ifndef __TIRE_PRESSURE_SENSOR_H__
+#define __TIRE_PRESSURE_SENSOR_H__
+
+#include "sensor.h"
```
Comment (ChedvaErblich) [inline control/include/tire_pressure_sensor.h]: We can use different sensors each time. How will this work? 
You have a header file for your example sensors, but what’s your plan for handling dynamic sensors? 
You’ll need to prepare the code for that.

and If it's only for testing, it should be placed in the test directory.

File: control/my_bson.bson
Comment (ChedvaErblich) [inline control/my_bson.bson:1]: Is this JSON for testing? It should be placed in the test folder.

File: control/src/basic_condition.cpp [hunk 19]
```diff
@@ -0,0 +1 @@
+#include "basic_condition.h"
```
Comment (ChedvaErblich) [inline control/src/basic_condition.cpp]: Consider deleting this file if it's not needed

File: control/src/enums.cpp [hunk 20]
```diff
@@ -0,0 +1,32 @@
+#include "enums.h"
+
+// Function that take a string and return the corresponding operator type
+OperatorTypes convertStringToOperatorTypes(string operatorType)
+{
+    // Define a map to convert from string to operator type
```
Comment (ChedvaErblich) [inline control/src/enums.cpp]: Consider refactoring the function to use a static unordered_map inside the function or as a static member of the class.
This approach ensures that the map is constructed only once and reused, which enhances the performance and efficiency of your code.

File: control/src/enums.cpp [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        {">=", OperatorTypes::be}, {"<=", OperatorTypes::le}};
+
+    // Find the right type in the map and return it
+    auto it = map.find(operatorType);
+    return ((it != map.end()) ? (it->second)
+                              : (OperatorTypes::UnknownOperatorType));
+}
+
+// Function that take a string and return the corresponding sensor type
+SensorsTypes convertStringToSensorsTypes(string type)
```
Comment (ChedvaErblich) [inline control/src/enums.cpp]: Again, you can't know the sensors types.
You will get them on run time.

File: control/src/main.cpp [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+#include "test_helper.h"
+using namespace std;
+
+int main()
+{
+    GlobalProperties &instanceGP = GlobalProperties::getInstance();
+    // Build the conditions from the bson file
+    Input::s_buildConditions();
+
+    vector<pair<int, string>> vec1 = {{5, "slow down"}, {8, "ccc"}};
```
Comment (ChedvaErblich) [inline control/src/main.cpp]: What is it?

File: control/src/main.cpp [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    vector<pair<int, string>> vec2 = {{8, "good!!!"}};
+
+    // Build the condition tree
+    FullCondition cond(
```
Comment (ChedvaErblich) [inline control/src/main.cpp]: The condition should be received via the bson file

File: control/src/main.cpp [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+        "|([5]&(|(=(code,500),<(status,\"high\")),=(msg,\"aaa\")),[5]&(|(=("
+        "code,500),<(status,\"high\")),>(msg,\"aaa\")))",
+        vec1);
+    instanceGP.conditions.insert({cond.id, &cond});
+    FullCondition c2("[5]|(=(code,500),<(status,\"high\"))", vec2);
+    instanceGP.conditions.insert({c2.id, &c2});
+
+    // --Test updates in the sensors--
```
Comment (ChedvaErblich) [inline control/src/main.cpp]: ?

File: control/src/main.cpp [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+    // Vectors for fields, values, and IDs to be updated
+    vector<string> fields = {"code", "code", "status", "msg", "status"};
+    vector<string> values = {"800", "500", "\"high\"", "\"aaa\"", "600"};
+    vector<int> ids = {8, 5, 5, 5, 8};
+
+    // Number of updates to process
+    size_t numUpdates = ids.size();
+
+    // Loop through each update
+    for (int i = 0; i < numUpdates; ++i) {
```
Comment (ChedvaErblich) [inline control/src/main.cpp]: we need an infinite loop that waits for input

File: control/src/global_properties.cpp [hunk 26]
```diff
@@ -0,0 +1,29 @@
+#include "global_properties.h"
+
+// Initializes the sensors based on a JSON file
+GlobalProperties::GlobalProperties()
+{
+    // Build the sensors according the json file
+    Input::s_buildSensors(sensors);
```
Comment (ChedvaErblich) [inline control/src/global_properties.cpp:51]: Can we use MIL here?

File: control/src/main.cpp
Comment (ChedvaErblich) [inline control/src/main.cpp:1]: The main function should be clean.
Don't put strings or other temporary values here for testing.

It should be ready for integration without temporary values. Instead, you can create fake JSON files to read from, but make sure to use the same behavior that will be used in the final version.

File: control/src/global_properties.cpp [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+// Resets the singleton instance
+void GlobalProperties::resetInstance()
+{
+    // Creates a new instance, replacing the existing one
+    instance.reset(new GlobalProperties());
+}
+
+// Static member to hold the single instance of `GlobalProperties`
+unique_ptr<GlobalProperties> GlobalProperties::instance = NULL;
```
Comment (ChedvaErblich) [inline control/src/global_properties.cpp]: Use nullptr

File: control/src/global_properties.cpp [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+    // Creates the instance if it does not exist
+    if (!instance) {
+        instance = unique_ptr<GlobalProperties>(new GlobalProperties());
+    }
+    // return Reference to the singleton `GlobalProperties` instance
+    return *instance;
+}
+
+// Resets the singleton instance
+void GlobalProperties::resetInstance()
```
Comment (ChedvaErblich) [inline control/src/global_properties.cpp:71]: Is it needed for the program?

The resetInstance method seems unusual for a singleton pattern, as singletons are typically designed to exist for the lifetime of the program. If you have a strong reason for "resetting" the instance, you can keep it, but typically singletons don't allow instance resets. Consider if this method is necessary.

I saw you use in for testing prוpose, but it is less common

File: control/src/input.cpp [hunk 29]
```diff
@@ -0,0 +1,241 @@
+#include "input.h"
```
Comment (ChedvaErblich) [inline control/src/input.cpp:1]: Add a blank line

File: control/src/input.cpp [hunk 30]
```diff
@@ -0,0 +1,241 @@
...
+                name = bson_iter_utf8(&docIter, NULL);  // Extract the sensor name/type.
+        }
+
+        // Convert the string to a sensor type enum.
+        SensorsTypes sensorType = convertStringToSensorsTypes(name);
+        Sensor *sensorPtr = nullptr;  // Pointer to the sensor object.
+        
+        // Create the appropriate sensor object based on the sensor type.
+        switch (sensorType) {
+            case SensorsTypes::Speed: {
```
Comment (ChedvaErblich) [inline control/src/input.cpp]: You don't know which sensors will be available, so why not always use the generic sensor?

File: control/src/input.cpp [hunk 31]
```diff
@@ -0,0 +1,241 @@
...
+                break;
+            }
+            default: {
+                sensorPtr = new Sensor(id);  // Create a generic Sensor.
+                break;
+            }
+        }
+
+        // Add the sensor pointer to the map if it was successfully created.
+        if (sensorPtr != nullptr)
```
Comment (ChedvaErblich) [inline control/src/input.cpp]: What will you do in case of error?

File: control/src/input.cpp [hunk 32]
```diff
@@ -0,0 +1,241 @@
...
+{
+    // Get a reference to the GlobalProperties singleton instance.
+    GlobalProperties &instanceGP = GlobalProperties::getInstance();
+
+    // Get "Conditions" array from the BSON document.
+    // If the "Conditions" array is not found or it is not valid, log an error and return.
+    bson_iter_t iter;
+    if (!bson_iter_init_find(&iter, document, "Conditions") || !BSON_ITER_HOLDS_ARRAY(&iter)) {
+        cerr << "Failed to find 'Conditions' array or it is not an array!" << endl;
+        return;
```
Comment (ChedvaErblich) [inline control/src/input.cpp:77]: Consider throwing an exception or returning an error code.

File: control/src/input.cpp [hunk 33]
```diff
@@ -0,0 +1,241 @@
...
+
+    // Create a BSON object from the extracted array data.
+    bson_t *bsonConditions = bson_new_from_data(conditionsData, conditionsLength);
+
+    // Initialize an iterator to iterate over the "Conditions" array.
+    bson_iter_t arrayIter;
+    if (!bson_iter_init(&arrayIter, bsonConditions)) {
+        // If initialization fails, clean up and return.
+        bson_destroy(bsonConditions);
+        return;
```
Comment (ChedvaErblich) [inline control/src/input.cpp:94]: Same here, you should implement a more robust error handling flow.

File: control/src/input.cpp [hunk 34]
```diff
@@ -0,0 +1,241 @@
...
+        // Clear the actions vector for the next iteration.
+        actions.clear();
+    }
+
+    // Clean up by destroying the BSON object.
+    bson_destroy(bsonConditions);
+}
+
+// Function that read the bson file
+bson_t *Input::s_readData()
```
Comment (ChedvaErblich) [inline control/src/input.cpp:195]: Same here, you should implement a more robust error handling flow.
for the whole function

File: control/src/operator_node.cpp [hunk 35]
```diff
@@ -0,0 +1,22 @@
...
+
+// Function that updates the status of all nodes that should change
+void OperatorNode::updateTree()
+{
+    bool prevStatus = this->status;
+
+    // Update the status of the current node
+    this->updateStatus();
+
+    // TODO: Check what would happen if the root was true and we got to it with true
```
Comment (ChedvaErblich) [inline control/src/operator_node.cpp:12]: Did you check it?
if it's ok now, please remove this comment

File: control/include/sensor.h [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+    // Contains the current values of various fields and a list of basic conditions associated with each field
+    map<string, pair<string, vector<BasicCondition *>>> fields;
+
+    // C-tor initializes the id member variable.
+    Sensor(int id) : id(id) {}
+
+    // Updates the condition status according to the received field and returns the  list of the full conditions whose root is true
+    void updateTrueRoots(string field, string value);
+    // Executes a specified action for the sensor.
+    virtual void doAction(string action) {}
```
Comment (ChedvaErblich) [inline control/include/sensor.h]: You should implement a generic doaction func based on the sensor ID or something else

File: control/src/test_helper.cpp
Comment (ChedvaErblich) [inline control/src/test_helper.cpp:1]: test should be located under the test dir.

**Comment by ChedvaErblich** [comment]
A few general comments:

1. Ensure comprehensive error handling is implemented throughout the code.
2. Since you can’t predict which sensors will be used, you should adopt a generic approach to handle them. And base it on the sensor IDs for better efficiency
3. Use references instead of making hard copies wherever possible to improve efficiency.
4. Avoid adding code or values solely for temporary testing purposes.
5. Keep runtime performance in mind and strive for efficiency in your implementations.

**Comment by MiryamBenHaim** [comment]
It is the first fixing, I will continue and send the second

**Comment by MiryamBenHaim** [comment]
This is the second fix

**Pull request closed without merging**